text/3512-closure-move-bindings.md

```diff
@@ -0,0 +1,281 @@
+the `IdentifierPattern` that declares the binding must have the `mut` keyword.
+
+If it is declared in an _UnnamedMoveBinding_,
+the `mut` keyword must be added in front of the expression;
+since the declared binding is always the first token in the expression,
+the `mut` token is always immediately followed by the mutable binding,
+thus yielding consistent readability.
+
+If it is implicitly captured from the parent scope
+instead of declared in a _MoveBinding_,
+the local variable declaration must be declared `mut` too.
+
+## _FullMove_ closures
+
+When the `move` keyword is present but _MoveBindings_ is absent (with its parentheses absent as well),
+the closure expression is of the _FullMove_ type, where
+all local variables in the closure construction scope
+are implicitly moved or copied into the closure on demand.
+
+Note that `move` with an empty pair of parentheses is allowed and follows the former rule;
+in other words, `move() |...| {...}` and `|...| {...}` are semantically equivalent.
+This allows macros to emit repeating groups of `_MoveBinding_ ","` inside a pair of parentheses
+and achieve correct semantics when there are zero repeating groups.
+
+If a moved binding is mutated inside the closure,
+its declaration in the parent scope must be declared `mut` too.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Due to backwards compatibility, this RFC proposes a new syntax
+that is an extension of capture-by-move
+but actually looks more similar to capture-by-reference,
+thus confusing new users.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Capture-by-reference is the default behavior for implicit captures for two reasons:
+
+1. It is more consistent to have `move(x)` imply `move(x=x)`,
+which leaves us with implicit references for the unspecified.
+2. Move bindings actually define a new shadowing binding
+that is completely independent of the original binding,
+so it is more correct to have the new binding explicitly named.
+Consider how unintuitive it is to require that
+a moved variable be declared `mut` in the outer scope
+even though it is only mutated inside the closure (as the new binding).
+
+The possible syntax for automatically-inferred _MoveBinding_ pattern
+is strictly limited to allow maximum future compatibility.
+Currently, many cases of captured bindings are in the form of
+`foo = foo`, `foo = &foo` or `foo.clone()`.
+This RFC intends to solve the ergonomic issues for these common scenarios first
+and leave more room for future enhancement when other frequent patterns are identified.
+
+Alternative approaches previously proposed
+include explicitly adding support for the `clone` keyword.
+This RFC does not favor such suggestions
+as they make the fundamental closure expression syntax
+unnecessarily dependent on the `clone` language item,
+and does not offer possibilities for alternative transformers.
+
```

Comment:
```suggestion

## `move` inside parameter list

As an alternative to having `move(binding)` ahead of a closure, we could put `move binding` in the parameter list: `|x: T, move y, move z = z.clone()| { ... }`.

This would have the advantage of keeping the list of declared names in one place, and giving a view of closures as having some bindings passed in as arguments and other bindings captured from the containing scope.

However, this would have the disadvantage of visually looking like the caller could pass the value in as a parameter, which it cannot.
```